shared/src/chain_state.rs

```diff
@@ -469,51 +476,68 @@ impl ChainState {
                     );
                 }
             } else {
-                tx_pool.enqueue_tx(entry.cycles, entry.size, entry.transaction);
+                let ret = self.pending_tx(tx_pool, entry.cycles, entry.size, entry.transaction);
+                if ret.is_err() {
+                    tx_pool.update_statics_for_remove_tx(entry.size, entry.cycles.unwrap_or(0));
+                    trace_target!(
+                        crate::LOG_TARGET_TX_POOL,
+                        "pending tx {:x} failed {:?}",
+                        tx_hash,
+                        ret
+                    );
+                }
             }
         }
     }
 
-    pub(crate) fn proposed_tx(
+    fn calculate_transaction_fee(&self, rtx: &ResolvedTransaction) -> Result<Capacity, PoolError> {
+        DaoCalculator::new(&self.consensus, self.store())
+            .transaction_fee(&rtx)
+            .map_err(|err| {
+                error_target!(
+                    crate::LOG_TARGET_TX_POOL,
+                    "Failed to generate tx fee for {:x}, reason: {:?}",
+                    rtx.transaction.hash(),
+                    err
+                );
+                PoolError::TxFee
+            })
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn send_tx_to_pool<F>(
         &self,
+        pool_name: &str,
         tx_pool: &mut TxPool,
         cycles: Option<Cycle>,
         size: usize,
         tx: Transaction,
-    ) -> Result<Cycle, PoolError> {
+        tx_resolved_result: Result<(Cycle, Capacity), PoolError>,
```

Comment:
I can't understand why we need `tx_resolved_result` here, what's the difference between fn param `cycles: Option<Cycle>` and `Result<Cycle>`, could you brief me?

Reply:
see the following code, we need to put tx into different pools by error type.